zebra: add kernel-bypass route map.

bigbanglwb opened the pull request:
Also provide UT in tests/topotests/kernel_bypass*

  detail: add a route map in zebra. When match satisfied, kernel bypass will set flag on route's nexthop.
  Flag will move to route entry, and finally move to zebra ctx as route data structure changed.
  I don't use DPLANE_CTX_FLAG_NO_KERNEL because this flag works bad.

ton31337 commented on tests/topotests/Dockerfile:
```diff
@@ -53,7 +53,16 @@ RUN export DEBIAN_FRONTEND=noninteractive \
     && echo "deb https://deb.frrouting.org/frr bionic frr-stable" > /etc/apt/sources.list.d/frr.list \
     && apt-get update \
-    && apt-get install -y libyang-dev \
+    && apt-get install -y \
```
This change should go into a separate commit.

bigbanglwb replied:
ok

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/.DS_Store:
Remove this file please.

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/r1/staticd.conf:
```diff
@@ -0,0 +1,10 @@
+# ip route 10.10.0.1/32 lo
```
Why is this file needed at all if all lines are commented?

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python
+
+#
```
Drop all the text and leave only Copyright (your) + SPDX-License-Identifier is a MUST (See other examples).

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+#
+
+"""
+<template>.py: Test <template>.
```
Useless comment.

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+    # daemon exists.
+    for router in router_list.values():
+        # Check for Version
+        if router.has_version('<', '5.1'):
```
Why do we need this check?

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+        expected = json.loads(open(ref_file).read())
+        test_func = partial(topotest.router_json_cmp,
+                            router, 'show bgp neighbors json', expected)
+        _, res = topotest.run_and_expect(test_func, None, count=10, wait=10.0)
```
Can we reduce the wait time to 1?

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+        expected = json.loads(open(ref_file).read())
+        test_func = partial(topotest.router_json_cmp,
+                            router, 'show ip route summary json', expected)
+        _, res = topotest.run_and_expect(test_func, None, count=10, wait=5)
```
Same here about wait time.

ton31337 commented on tests/topotests/kernel_bypass_zebra_ecmp/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,168 @@
+        pytest.skip(tgen.errors)
+
+    r1 = tgen.gears['r1']
+    test_func = partial(r1.cmd,"route -n | wc -l")
```
Please use `ip route` instead. Also, it would be easier to check with `ip -j route`.

bigbanglwb replied:
ip -j route is not supported in low kernel version, like my testbed of ubuntu 18.04~4.15.0-193-generic.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/.DS_Store:
Drop this file, please.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/r1/staticd.conf:
```diff
@@ -0,0 +1,10 @@
+# ip route 10.10.0.1/32 lo
```
Useless file.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/r1/zebra.conf:
```diff
@@ -0,0 +1,10 @@
+interface r1-eth1
+ ip address 192.168.1.1/24
+
+# debug zebra kernel
```
Use ! for commenting the lines.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+#!/usr/bin/env python
+
+#
+# <template>.py
```
Ditto regarding Copyright, SPDX-License-Identifier, etc. See previous comment.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+#
+
+"""
+<template>.py: Test <template>.
```
Ditto.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+import json
+import re
+from functools import partial
+import pdb
```
Where is this used?

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+import sys
+import pytest
+import json
+import re
```
Where is this used?

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+    # daemon exists.
+    for router in router_list.values():
+        # Check for Version
+        if router.has_version('<', '5.1'):
```
Ditto.

ton31337 commented on tests/topotests/kernel_bypass_zebra_reconfig/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,197 @@
+        pytest.skip(tgen.errors)
+
+    r1 = tgen.gears['r1']
+    test_func = partial(r1.cmd,"route -n | wc -l")
```
Ditto.

ton31337 commented on zebra/rt_netlink.c:
```diff
@@ -2969,13 +2969,29 @@ netlink_put_route_update_msg(struct nl_batch *bth, struct zebra_dplane_ctx *ctx)
+	flag = dplane_ctx_get_flag(ctx);
+	old_flag = dplane_ctx_get_old_flag(ctx);
+	if (IS_ZEBRA_DEBUG_KERNEL)
+		zlog_debug("%s:rt_netlink, route op = %s,route perfix: %pFX,zebra ctx flag = 0x%x, old flag = 0x%x",
```
typo perfix, whitespace is missing before `route` and `zebra`.

ton31337 commented on zebra/zebra_dplane.c:
```diff
@@ -4187,7 +4223,7 @@ dplane_route_update_internal(struct route_node *rn,
 			== dplane_ctx_get_old_nhe_id(ctx))
 		    && (dplane_ctx_get_nhe_id(ctx) >= ZEBRA_NHG_PROTO_LOWER)) {
-			struct nexthop *nexthop;
+			// struct nexthop *nexthop;
```
If not needed, drop it at all then.

ton31337 commented on zebra/zebra_dplane.h:
```diff
@@ -53,6 +53,7 @@ zebra_dplane_info_from_zns(struct zebra_dplane_info *zns_info,
 #endif /* NETLINK */
 }
 
+
```
Drop this change.

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -640,6 +640,42 @@ DEFPY_YANG(
+	    set_kernel_bypass_cmd,
+	    "set kernel-bypass",
+	    SET_STR
+	    "kernel bypass\n")
```
Very short explanation.. Please imagine yourself as an operator and try to understand what it means.

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -640,6 +640,42 @@ DEFPY_YANG(
+	    SET_STR
+	    "kernel bypass\n")
+{
+	zlog_debug("zebra set route map rule : kernel-bypass ");
```
Unguarded debug log, please guard it.

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -640,6 +640,42 @@ DEFPY_YANG(
+	return nb_cli_apply_changes(vty, NULL);
+}
+
+DEFPY_YANG(no_set_kernel_bypass,
```
Combine this command with the positive form.

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -1549,6 +1585,59 @@ static const struct route_map_rule_cmd route_set_src_cmd = {
+				break;
+			case NEXTHOP_TYPE_IPV6_IFINDEX:
+			case NEXTHOP_TYPE_IPV6:
+				inet_ntop(AF_INET6, &nh_data->nexthop->gate.ipv6, ipv6_str, INET6_ADDRSTRLEN);
```
Why can't we print a string instead (ipv6_str) without converting?

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -1549,6 +1585,59 @@ static const struct route_map_rule_cmd route_set_src_cmd = {
+	return (void*) 1;
+}
+
+static void route_set_kernel_bypass_free(void *rule)
```
Use just route_value_free().

bigbanglwb replied:
static void route_set_kernel_bypass_free(void *rule)
{
	XFREE(MTYPE_ROUTE_MAP_COMPILED, rule);;
}

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -1549,6 +1585,59 @@ static const struct route_map_rule_cmd route_set_src_cmd = {
+}
+static void *route_set_kernel_bypass_compile(const char *arg)
+{
+	return (void*) 1;
```
`return XSTRDUP(MTYPE_ROUTE_MAP_COMPILED, arg);` - to be consistent with others.

bigbanglwb replied:
I learnt this return way from route-map "atomic-aggregate". Actually when I use `XSTRDUP`, it return NULL. Because this route map will be like "set kernel-bypass", which means "arg" of `route_set_kernel_bypass_compile` is NULL.

ton31337 reviewed (changes requested):
* Documentation is missing
* Fix ![1691572583](https://github.com/FRRouting/frr/assets/3352707/8e30c34a-9266-4529-afc9-6b33320ec48f) for all files it's reported as a warning
* Lots of inconsistent grammar stuff (especially for logging) - keep whitespace after `:`, `,`, etc.

ton31337 commented:
Also, styling fixes MUST be applied (see frrbot check).

mjstapp reviewed (commented):
I'd like more description of what you're trying to achieve here. You're sort of passing flag bits around among various places, but it's hard to tell exactly why.
I'd also like to know why the existing "NO_KERNEL" mechanism doesn't work for you - I'd prefer to fix or extend something that we have already if possible rather than add a second, parallel version?

donaldsharp reviewed (changes requested):
This commit must be broken up into smaller commits.  It is too large and the functionality will be impossible to access

donaldsharp commented on zebra/rt_netlink.c:
```diff
@@ -2969,13 +2969,29 @@ netlink_put_route_update_msg(struct nl_batch *bth, struct zebra_dplane_ctx *ctx)
 
-		if (p->family == AF_INET || v6_rr_semantics) {
+		if ((p->family == AF_INET || v6_rr_semantics) && 
+			(!RSYSTEM_ROUTE(dplane_ctx_get_type(ctx))) &&
```
The check of the flag should just be the first check( line 2972/2985).

donaldsharp commented on zebra/rt_socket.c:
```diff
@@ -312,28 +312,32 @@ enum zebra_dplane_result kernel_route_update(struct zebra_dplane_ctx *ctx)
+	frr_with_privs(&zserv_privs) {
 		if (dplane_ctx_get_op(ctx) == DPLANE_OP_ROUTE_DELETE) {
-			if (!RSYSTEM_ROUTE(type))
+			if (!RSYSTEM_ROUTE(type) && !CHECK_FLAG(flag, DPLANE_RINFO_FLAG_NO_KERNEL))
```
again why not just do this first? instead of having to have a check flag call in every possible combination

donaldsharp commented on zebra/zebra_dplane.c:
```diff
@@ -4132,19 +4149,38 @@ dplane_route_update_internal(struct route_node *rn,
+	struct nexthop *nexthop, *old_nexthop;
 	/* Obtain context block */
 	ctx = dplane_ctx_alloc();
+	if (IS_ZEBRA_DEBUG_KERNEL) {
```
How does this debug help the end developer debug this.

bigbanglwb replied:
Just for me to observe flags' change. Should I delete them?

donaldsharp reviewed (changes requested):
see comments but lots of cleanup is needed

bigbanglwb commented:
@mjstapp There is a scenario where I dynamically mount a route-map, and routes will marked as kernel bypass. But these routes have been installed in kernel, I need to delete these routes. The original skip_kernel does not support this requirement. I am unsure if it would impact other users while I modify skip_kernel, so I just add.

bigbanglwb commented on lib/nexthop.h:
```diff
@@ -81,6 +81,7 @@ struct nexthop {
 #define NEXTHOP_FLAG_SRTE       (1 << 7) /* SR-TE color used for BGP traffic */
 #define NEXTHOP_FLAG_EVPN       (1 << 8) /* nexthop is EVPN */
 #define NEXTHOP_FLAG_LINKDOWN   (1 << 9) /* is not removed on link down */
+#define NEXTHOP_FLAG_KERNEL_BYPASS   (1 << 10) /* nexthop is set to bypass kernel deliver*/
```
for route map, function can only get nexthop, I used it for save flag temporarily.
![image](https://github.com/FRRouting/frr/assets/43382319/599e24c5-b641-44ec-8119-218a50be3e7e)

donaldsharp commented:
https://github.com/FRRouting/frr/pull/14183 -> Please use this as a foundation for the route map matching and create a match that is not stored in the nexthop, it is stored as a flag on the  route entry

bigbanglwb commented:
> * Documentation is missing
> * Fix ![1691572583](https://user-images.githubusercontent.com/3352707/259363555-8e30c34a-9266-4529-afc9-6b33320ec48f.png) for all files it's reported as a warning
> * Lots of inconsistent grammar stuff (especially for logging) - keep whitespace after `:`, `,`, etc.

![image](https://github.com/FRRouting/frr/assets/43382319/235082bd-d314-4570-9bc3-8ec8f7d788b9)
Where can I find the cause of these files' warning ?

bigbanglwb commented:
> 

ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

ton31337 commented on lib/routemap_cli.c:
```diff
@@ -1358,6 +1358,11 @@ void route_map_action_show(struct vty *vty, const struct lyd_node *dnode,
 				dnode,
 				"./rmap-set-action/frr-bgp-route-map:l3vpn-nexthop-encapsulation"));
+	} else if (IS_SET_ZEBRA_KERNEL_BYPASS(action)) {
+		vty_out(vty, " set route kernel-bypass %s\n",
```
`set route kernel-bypass`, but in documentation we have `set kernel-bypass`. Which one is correct?

bigbanglwb commented:
ci:rerun

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/r2/staticd.conf:
```diff
@@ -0,0 +1,11 @@
+ip route 20.30.0.1/32 r2-eth0
```
Please use https://datatracker.ietf.org/doc/html/rfc5737 for IPv4 addresses.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/r3/staticd.conf:
```diff
@@ -0,0 +1,10 @@
+ip route 20.30.0.1/32 r3-eth0
```
Ditto.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,143 @@
+# SPDX-License-Identifier: ISC
+#
+# test_kernel_bypass_ecmp.py
+# Part of NetDEF Topology Tests
```
This header is not needed.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,143 @@
+# test_kernel_bypass_ecmp.py
+# Part of NetDEF Topology Tests
+#
+# Copyright (c) 2017 by
```
This copyright is not needed as well. Add yours.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/python
```
`pytestmark` is missing, see other examples.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,143 @@
+    for rname, router in router_list.items():
+        if rname == "r1":
+            router.load_frr_config(
+                os.path.join(CWD, "{}/frr_routemap.conf".format(rname))
```
Why do we need two files frr.conf and frr_routemap.conf for r1?

bigbanglwb replied:
I want to use 2 frr.conf for 2 test "test_ecmp", "test_reconfig" before. 
"test_ecmp" is to test route-map is configured before zebra init and "test_ecmp" is to test route-map is configured after zebra init.
I find  "test_reconfig" can cover "test ecmp" normal case now, I will remove  "test_ecmp" now.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_ecmp.py:
```diff
@@ -0,0 +1,143 @@
+    route_connected = 2
+
+    r1 = tgen.gears['r1']
+    test_func = partial(r1.cmd, "ip route  | wc -l")
```
Please use `ip route -j` (JSON parsing).

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/python
+# SPDX-License-Identifier: ISC
+#
+# test_kernel_bypass_reconfig.py
```
Same as before regarding copyrights, etc.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,168 @@
+    route_connected = 2
+
+    r1 = tgen.gears['r1']
+    test_func = partial(r1.cmd, "ip route  | wc -l")
```
Ditto.

ton31337 commented on tests/topotests/kernel_bypass_route_map_zebra/test_kernel_bypass_reconfig.py:
```diff
@@ -0,0 +1,168 @@
+        """
+    )
+
+    test_func = partial(r1.cmd, "ip route | wc -l")
```
Ditto.

ton31337 commented on zebra/rt_socket.c:
```diff
@@ -312,7 +313,21 @@ enum zebra_dplane_result kernel_route_update(struct zebra_dplane_ctx *ctx)
+	flag = dplane_ctx_get_flags(ctx);
+	old_flag = dplane_ctx_get_old_flags(ctx);
 	frr_with_privs(&zserv_privs) {
+		/* If new route is set kernel-bypass,we just return sunccess
```
typo sunccess

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -636,6 +636,32 @@ DEFPY_YANG(
+	char xpath_value[XPATH_MAXLEN];
+
+	if (IS_ZEBRA_DEBUG_KERNEL)
+		zlog_debug("zebra set route map rule: kernel-bypass");
```
I don't understand why this debug is so useful?

ton31337 commented on zebra/zebra_routemap.c:
```diff
@@ -636,6 +636,32 @@ DEFPY_YANG(
+
+	if (IS_ZEBRA_DEBUG_KERNEL)
+		zlog_debug("zebra set route map rule: kernel-bypass");
+	if (strmatch(argv[0]->text, "no"))
```
Use just `no`, no need to parse the text:

```c
if (no)
```

Commit:
test: add kernel_bypass route-map topotest

  test_kernel_bypass_reconfig: test dynamically configure
  kernel_bypass route_map, can cover normal case.

Signed-off-by: 乐倚 <lwb406054@alibaba-inc.com>

Commit:
doc: add kernel-bypass route-map doc

Signed-off-by: 乐倚 <lwb406054@alibaba-inc.com>

Commit:
cli: add kenel-bypass route-map cli

Signed-off-by: 乐倚 <lwb406054@alibaba-inc.com>

Commit:
zebra: add kernel_bypass route map

  Add flag 'ZEBRA_FLAG_KERNEL_BYPASS' in lib/zclient.c
  to instuct kernel bypass.When match  statement satisfied,
  route map will set flag on re flags. Kernel bypass flag will
  move to zebra ctx in zebra.

Signed-off-by: 乐倚 <lwb406054@alibaba-inc.com>

Commit:
zebra: handle kernel bypass flag

  Handle route marked as kernel bypass. Zebra will convert
  flag to zebra ctx, and skip route install to kernel.

Signed-off-by: 乐倚 <lwb406054@alibaba-inc.com>

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

bigbanglwb commented:
ci:rerun

riw777 commented:
@frrbot autoclose in 2 weeks
